File: backend/app/models/asset.py
from __future__ import annotations
from pydantic import BaseModel
from pydantic.generics import GenericModel
from typing import Generic, Literal, Optional, TypeVar

T = TypeVar("T", bound=BaseModel)


class Asset(GenericModel, Generic[T]):
  name: str
  id: int
  type: Literal["folder", "file"]
  preview: Optional[int] = None
  parent_id: int = 0
  data: Optional[T] = None

File: frontend/types/schemas/project/Asset.ts
import { z } from "zod";

export const AssetSchema = z.object({
  name: z.string(),
  id: z.number().int(),
  type: z.enum(["folder", "file"]),
  preview: z.union([z.number().int(), z.null()]).default(null),
  parent_id: z.number().int().default(0),
  data: z.union([z.object({}), z.null()]).default(null),
});
export type Asset = z.infer<typeof AssetSchema>;
